src/Mod/Sketcher/App/Sketch.cpp

```diff
@@ -1704,11 +1717,30 @@ int Sketch::addConstraint(const Constraint *constraint)
         break;
     case Tangent:
         if (constraint->FirstPos == PointPos::none &&
-                constraint->SecondPos == PointPos::none &&
-                constraint->Third == GeoEnum::GeoUndef){
+            constraint->SecondPos == PointPos::none &&
+            constraint->Third == GeoEnum::GeoUndef){
             //simple tangency
             rtn = addTangentConstraint(constraint->First,constraint->Second);
         }
+        else if (constraint->FirstPos == PointPos::start &&
+                 constraint->SecondPos == PointPos::none &&
+                 constraint->Third == GeoEnum::GeoUndef) {
+            // check for B-Spline Knot to curve tangency
+            auto knotgeoId = checkGeoId(constraint->First);
+            if (Geoms[knotgeoId].type == Point) {
+                auto *point = static_cast<const GeomPoint*>(Geoms[knotgeoId].geo);
+
+                if (GeometryFacade::isInternalType(point,InternalType::BSplineKnotPoint)) {
+                    auto bsplinegeoid = internalAlignmentGeometryMap.at(constraint->First);

```

Comment:
I think here we have an extra check missing for a line. If the curve is not a line and the point is not a knot (added for example from Python), then this should return -1, as it is an unsupported constraint (malformed constraint).

Reply:
I think there are quite a few ways that we can possibly make malformed constraints. For example perpendiculars should now cause them too. And as we discussed earlier the 3-item tangent via point dors pass the UI. Any solution to this without looking at the entire intent will probably just be patchwork.

Reply:
PS: note I haven't tried the behaviour for perpendiculars yet.

Reply:
The UI does not allow a perpendicular to be set. This code is specific for adding tangencies. Not sure how it could affect perpendiculars. 

In fact, that is an advantage of not using the 3 point framework for non AngleViaPoint solver implementations. As the 3 point framework was designed for AngleViaPoint and AngleViaPoint code is generic to seamlessly support tangencies and perpendiculars. But the present constraint is only for tangency.

I think the framework is very general to allow anything that fits (and for which we could eventually provide support). However, this necessarily requires that the unsupported cases are handled by the software. Because we have two ways to work (UI and Python), and the level of interaction with the user is certainly different (as Python calls are prominently from macros), we are forced to check it first at UI level to provide the interactive user with the relevant error, and notify the malformed constraint for macros using Python. I do not really see how this could be better streamlined...

Reply:
OK just checked again and perpendiculars are indeed blocked (though it should be straightforward to implement it). I thought I had changed `isSimplePoint` but I hadn't and had just added a knot check.

Reply:
Just put the checks, both in GUI and solver-interface side.